#181: Documenting type parameter modes and fixing the check of !new (#941)
dafny-lang/dafny: docs/_includes/Types.md

  "<" TypeVariableName [ "(" "==" ")" ]
  { "," TypeVariableName [ "(" "==" ")" ] } ">"
````

> Many of the types, functions and methods in Dafny can be

Reviewer: ```suggestion
Many of the types, functions, and methods in Dafny can be
```

Author: Usually, I'm the pedantic one on commas

types when used in a non-ghost context.
Again, note that _all_ types support equality in _ghost_
contexts; the difference is only for non-ghost (that is, compiled)
code.  Co-inductive datatypes, function types, and inductive
datatypes with ghost parameters are examples of types that are not
equality supporting.

### Auto-initializable types: `T(0)`

All Dafny variables of a given type hold a legal value of that type;
if no explicit initialization is given, then an arbitrary value is
assumed by the verifier and supplied by the compiler,
 that is, the variable is _auto-initialized_.
During verification, this means that any subsequent uses of that
variable must hold for any value.
For example,
```dafny
method m() {
  var n: nat; // Initialized to arbitrary value
  assert n >= 0; // true, regardless of the value of n
  var i: int;
  assert i >= 0; // possibly false, arbitrary ints may be negative
}
```

However, by default, non-ghost computations must follow definite-assignment
rules, in which case local variables must be assigned a value before use.
For some types, the compiler can choose an initial value, but for others

> it does not; for more details see the section describing [Definite Assignment](#sec-definite-assignment) and the `-definiteAssignment` command-line option.

Reviewer: The first sentence here is not right (perhaps because the `/help` message is not clear enough). I suggest replacing both sentences with:

For some types, the compiler can choose an initial value, but for others it does not.
Variables and fields of a type that the compiler does not auto-initialize are subject to
_definite-assignment rules_. These ensure that the program explicitly assigns the
variable before it is used. For more details, see the section describing
[Definite Assignment](#sec-definite-assignment) and the `-definiteAssignment` command-line
option.

Also, the `-help` description of `-definiteAssignment:1` can be clarified to:

    1 (default) - enforces definite-assignment rules for variables and fields
                  of types that do not support auto-initialization

Author: Done

The `(0)` suffix indicates that the type must be one that the compiler knows
how to auto-initialize, if the type is used to declare a non-ghost variable.

### Non-heap based: `T(!new)`

Dafny makes a distinction between types whose values are on the heap,
i.e. references, like
classes and arrays, and those that are strictly value-based, like basic
types and datatypes.
The practical implication is that references depend on allocation state
(e.g. are affected by the `old` operation) whereas non-reference values
are not.
Thus it can be relevant to know whether the values of a type parameter
are heap-based or not. This is indicated by the mode suffix `(!new)`.

A type parameter characterized by `(!new)` is _recursively_ independent

> of the allocation state. For example a datatype is not a reference, but for

Author: OK

declared inside angle brackets and can stand for any type.

It is sometimes necessary to restrict these type parameters so that
they can only be instantiated by certain families of types, that is,
by types that have certain properties. The following subsections
describe the restrictions Dafny supports.

Dafny has some inference support that makes certain signatures less
cluttered (described in Section [Type Inference](#sec-type-inference)).
In some cases, this support will infer that a type-parameter
must be restricted in a particular way, in which case Dafny
will add the appropriate suffix, such as `(==)`, automatically.

## Equality-supporting type parameters: `T(==)`

Designating a type parameter with the `(==)` suffix indicates that
the parameter may only be replaced with types that are known to

> support equality comparisons (`==` and `!=`).

Reviewer: ```suggestion
support run-time equality comparisons (`==` and `!=`).
```

Author: OK

This restriction only applies during compilation; for verification

> the `(==)` suffix does not restrict the substituted type.

Reviewer: If a type parameter is marked with `(==)`, it should always be restricted as noted in the previous sentence. But note that all types support equality in ghost contexts.

Reviewer: Oh, I just tried this and it current behaves as you described! I think that behavior is not desirable. That is, it should be that "this characteristic of a type parameter is operative for both verification and compilation", even ghost contexts never inspect this type characteristic.

Author: Fixed per discussion

datatypes with ghost parameters are examples of types that are not
equality supporting.

## Auto-initializable types: `T(0)`

All Dafny variables of a given type hold a legal value of that type;
if no explicit initialization is given, then an arbitrary value is

> assumed.

Reviewer: This could possible say "assumed by the verifier and provided by the compiler".

Author: OK

datatypes with ghost parameters are examples of types that are not
equality supporting.

### Auto-initializable types: `T(0)`

All Dafny variables of a given type hold a legal value of that type;
if no explicit initialization is given, then an arbitrary value is

> assumedi, that is, the variable is _auto-innitialized_..

Reviewer: ```suggestion
assumed, that is, the variable is _auto-initialized_.
```

Author: Fixed this along with other edits here
